py/server/deephaven/table_listener.py

```diff
@@ -72,208 +257,164 @@ def _do_locked(f: Callable, lock_type="shared") -> None:
     Raises:
         ValueError
     """
-    throwing_runnable = jpy.get_type(
-        "io.deephaven.integrations.python.PythonThrowingRunnable"
-    )
-    update_graph_processor = jpy.get_type(
-        "io.deephaven.engine.updategraph.UpdateGraphProcessor"
-    )
+    throwing_runnable = jpy.get_type("io.deephaven.integrations.python.PythonThrowingRunnable")
+    update_graph_processor = jpy.get_type("io.deephaven.engine.updategraph.UpdateGraphProcessor")
 
     if lock_type == "exclusive":
         update_graph_processor.DEFAULT.exclusiveLock().doLocked(throwing_runnable(f))
     elif lock_type == "shared":
+        # slock = update_graph_processor.DEFAULT.sharedLock()
+        # try:
+        #     slock.lock()
+        #     f()
+        # finally:
+        #     slock.unlock()
         update_graph_processor.DEFAULT.sharedLock().doLocked(throwing_runnable(f))
     else:
         raise ValueError(f"Unsupported lock type: lock_type={lock_type}")
 
 
-def _nargs_listener(listener) -> int:
-    """Returns the number of arguments the listener takes.
-
-    Args:
-        listener: listener
-
-    Returns:
-        the number of arguments the listener takes
+class TableListener(ABC):
+    """An abstract table listener class that should be subclassed by any user table listener class."""
 
-    Raises:
-        ValueError, NotImplementedError
-    """
+    @abstractmethod
+    def on_update(self, update: TableUpdate, is_replay: bool) -> None:
+        ...
 
-    if callable(listener):
-        f = listener
-    elif hasattr(listener, "onUpdate"):
-        f = listener.onUpdate
-    else:
-        raise ValueError(
-            "ShiftObliviousListener is neither callable nor has an 'onUpdate' method"
-        )
 
-    return len(signature(f).parameters)
+def listener_wrapper(table: Table):
+    def decorator(listener: Callable):
+        @wraps(listener)
+        def wrapper(update, *args):
+            sig = inspect.signature(listener)
+            n_params = len(sig.parameters)
+            t_update = TableUpdate(table=table, j_table_update=update)
 
+            if n_params == 2:
+                listener(t_update, args[0])
+            else:
+                listener(t_update)
```

Comment:
This looks problematic.
1) Is the 1 arg supported?
2) What if there are 3 or more args?  Shouldn't it raise an error?

Reply:
I see that functions using this are testing the 2 arg case.  The logic is spread out in multiple places.  If the logic was consolidated, it would be more simple.